File: sdks/js/src/__tests__/client.test.ts
import { describe, it, expect, vi, afterEach } from 'vitest';
import { SuppuoClient, SuppuoError } from '../index.js';

function envelope(data: unknown) {
  return {
    data,
    error: null,
    meta: { requestId: 'req_test', timestamp: new Date().toISOString() },
  };
}

function mockFetch(status: number, body: unknown) {
  return vi.fn(async () => new Response(JSON.stringify(body), { status })) as unknown as typeof fetch;
}

afterEach(() => {
  vi.unstubAllGlobals();
});

describe('SuppuoClient', () => {
  it('constructs with default base URL', () => {
    const c = new SuppuoClient({ token: 'test' });
    expect(c).toBeDefined();
  });

  it('lists tickets with Bearer auth and query params', async () => {
    const fetchSpy = mockFetch(200, envelope({ tickets: [], counts: {} }));
    vi.stubGlobal('fetch', fetchSpy);
    const c = new SuppuoClient({ token: 'tok_123', baseUrl: 'https://suppuo.test' });
    const out = await c.tickets.list({ status: 'open', limit: 10 });
    expect(out).toEqual({ tickets: [], counts: {} });
    const [url, init] = (fetchSpy as ReturnType<typeof vi.fn>).mock.calls[0]!;
    expect(String(url)).toBe('https://suppuo.test/api/v1/tickets?status=open&limit=10');
    expect((init as RequestInit).headers).toMatchObject({ Authorization: 'Bearer tok_123' });
  });

  it('unwraps envelope errors into SuppuoError with the code', async () => {
    vi.stubGlobal(
      'fetch',
      mockFetch(404, {
        data: null,
        error: { code: 'NOT_FOUND', message: 'ticket not found' },
        meta: { requestId: 'req_x', timestamp: 'now' },
      }),
    );
    const c = new SuppuoClient({ token: 'tok' });
    const err = await c.tickets.get('tkt_missing').catch((e: unknown) => e);
    expect(err).toBeInstanceOf(SuppuoError);
    expect((err as SuppuoError).code).toBe('NOT_FOUND');
    expect((err as SuppuoError).status).toBe(404);
    expect((err as SuppuoError).requestId).toBe('req_x');
  });

  it('public endpoints work without a token', async () => {
    const fetchSpy = mockFetch(201, envelope({ number: 1, accessToken: 'at_x' }));
    vi.stubGlobal('fetch', fetchSpy);
    const prev = process.env.SUPPUO_TOKEN;
    delete process.env.SUPPUO_TOKEN;
    try {
      const c = new SuppuoClient();
      const out = await c.public.submitTicket({
        accountId: 'acc_0123456789abcdef01234567',
        subject: 'Help',
        body: 'It broke',
        email: 'user@example.com',
      });
      expect(out.accessToken).toBe('at_x');
      const [, init] = (fetchSpy as ReturnType<typeof vi.fn>).mock.calls[0]!;
      expect((init as RequestInit).headers).not.toHaveProperty('Authorization');
    } finally {
      if (prev !== undefined) process.env.SUPPUO_TOKEN = prev;
    }
  });

  it('authed endpoints without a token fail fast with AUTH_REQUIRED', async () => {
    const prev = process.env.SUPPUO_TOKEN;
    delete process.env.SUPPUO_TOKEN;
    try {
      const c = new SuppuoClient();
      const err = await c.cannedReplies.list().catch((e: unknown) => e);
      expect((err as SuppuoError).code).toBe('AUTH_REQUIRED');
    } finally {
      if (prev !== undefined) process.env.SUPPUO_TOKEN = prev;
    }
  });
});

File: sdks/go/client_test.go
package suppuo

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestNew(t *testing.T) {
	c := New(Config{Token: "test"})
	if c == nil {
		t.Fatal("client is nil")
	}
	if c.Tickets == nil || c.CannedReplies == nil || c.Public == nil {
		t.Fatal("resource namespaces not wired")
	}
}

func TestTicketsListUnwrapsEnvelope(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/tickets" {
			t.Errorf("path = %s", r.URL.Path)
		}
		if got := r.Header.Get("Authorization"); got != "Bearer tok_123" {
			t.Errorf("auth header = %q", got)
		}
		if got := r.URL.Query().Get("status"); got != "open" {
			t.Errorf("status = %q", got)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"data":  map[string]any{"tickets": []any{}, "counts": map[string]int{"open": 2}},
			"error": nil,
			"meta":  map[string]any{"requestId": "req_t", "timestamp": "now"},
		})
	}))
	defer srv.Close()

	c := New(Config{Token: "tok_123", BaseURL: srv.URL})
	out, err := c.Tickets.List(context.Background(), &TicketListParams{Status: "open", Limit: 10})
	if err != nil {
		t.Fatal(err)
	}
	if out.Counts["open"] != 2 {
		t.Errorf("counts = %v", out.Counts)
	}
}

func TestEnvelopeErrorBecomesTypedError(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusNotFound)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"data":  nil,
			"error": map[string]any{"code": "NOT_FOUND", "message": "ticket not found"},
			"meta":  map[string]any{"requestId": "req_x", "timestamp": "now"},
		})
	}))
	defer srv.Close()

	c := New(Config{Token: "tok", BaseURL: srv.URL})
	_, err := c.Tickets.Get(context.Background(), "tkt_missing")
	var apiErr *Error
	if !errors.As(err, &apiErr) {
		t.Fatalf("expected *Error, got %T", err)
	}
	if apiErr.Code != "NOT_FOUND" || apiErr.Status != 404 || apiErr.RequestID != "req_x" {
		t.Errorf("unexpected error: %+v", apiErr)
	}
}

func TestPublicNeedsNoToken(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if got := r.Header.Get("Authorization"); got != "" {
			t.Errorf("unexpected auth header %q", got)
		}
		w.WriteHeader(http.StatusCreated)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"data":  map[string]any{"number": 1, "accessToken": "at_x"},
			"error": nil,
			"meta":  map[string]any{"requestId": "req_p", "timestamp": "now"},
		})
	}))
	defer srv.Close()

	c := New(Config{BaseURL: srv.URL, Token: ""})
	out, err := c.Public.SubmitTicket(context.Background(), SubmitTicketInput{
		AccountID: "acc_x", Subject: "s", Body: "b", Email: "e@example.com",
	})
	if err != nil {
		t.Fatal(err)
	}
	if out.AccessToken != "at_x" {
		t.Errorf("accessToken = %q", out.AccessToken)
	}
}

func TestAuthedWithoutTokenFailsFast(t *testing.T) {
	t.Setenv("SUPPUO_TOKEN", "")
	c := New(Config{BaseURL: "http://127.0.0.1:0"})
	_, err := c.CannedReplies.List(context.Background())
	var apiErr *Error
	if !errors.As(err, &apiErr) || apiErr.Code != "AUTH_REQUIRED" {
		t.Fatalf("expected AUTH_REQUIRED, got %v", err)
	}
}
